pkg/expr/convert_to_long.go

```diff
@@ -163,108 +163,139 @@ func convertNumericWideToNumericLong(frames data.Frames) (data.Frames, error) {
-						outputField.Set(idx, val)
-					}
+			if !field.Type().Numeric() {
+				continue
+			}
+
+			valueFieldNames[field.Name] = struct{}{}
+			for k := range field.Labels {
+				labelKeySet[k] = struct{}{}
+			}
 
-					// Add labels for the numeric field
-					for key, value := range field.Labels {
-						if outputField, exists := outputLabelFields[key]; exists {
-							outputField.Set(idx, value)
-						}
+			for i := 0; i < field.Len(); i++ {
+				t := timeField.At(i).(time.Time)
+				v, err := field.FloatAt(i)
+				if err != nil {
+					v = 0
+				}
+
+				key := rowKey(fmt.Sprintf("%d|%s", t.UnixNano(), field.Labels.Fingerprint()))
+				entry, exists := rowMap[key]
+				if !exists {
+					entry = rowEntry{
+						t:      t,
+						labels: field.Labels,
+						values: map[string]float64{},
 					}
 				}
+				entry.values[field.Name] = v
+				rowMap[key] = entry
 			}
 		}
 	}
 
-	// Build the output frame
-	outputFields := []*data.Field{timeField}
-	for _, field := range outputNumericFields {
-		outputFields = append(outputFields, field)
+	// Extract and sort row entries
+	type sortableRow struct {
+		key   rowKey
+		entry rowEntry
 	}
-	for _, field := range outputLabelFields {
-		outputFields = append(outputFields, field)
+	var sortedRows []sortableRow
+	for k, v := range rowMap {
+		sortedRows = append(sortedRows, sortableRow{key: k, entry: v})
 	}
-	outputFrame := data.NewFrame("time_series_long", outputFields...)
 
-	// Set metadata
-	if outputFrame.Meta == nil {
-		outputFrame.Meta = &data.FrameMeta{}
+	// Sort by time asc, then label fingerprint
+	sort.Slice(sortedRows, func(i, j int) bool {
+		ti := sortedRows[i].entry.t
+		tj := sortedRows[j].entry.t
+		if ti.Equal(tj) {
+			return sortedRows[i].entry.labels.Fingerprint().String() < sortedRows[j].entry.labels.Fingerprint().String()
+		}
+		return ti.Before(tj)
+	})
+
+	// Collect ordered value and label field names
+	valueNames := make([]string, 0, len(valueFieldNames))
+	for name := range valueFieldNames {
+		valueNames = append(valueNames, name)
+	}
+	sort.Strings(valueNames)
+
+	labelKeys := make([]string, 0, len(labelKeySet))
+	for k := range labelKeySet {
+		labelKeys = append(labelKeys, k)
+	}
+	sort.Strings(labelKeys)
+
+	// Build columns
+	numRows := len(sortedRows)
+	timeCol := make([]time.Time, numRows)
+	valueCols := make(map[string][]float64, len(valueNames))
+	for _, name := range valueNames {
+		valueCols[name] = make([]float64, numRows)
+	}
+	labelCols := make(map[string][]string, len(labelKeys))
+	for _, k := range labelKeys {
+		labelCols[k] = make([]string, numRows)
 	}
-	outputFrame.Meta.Type = data.FrameTypeTimeSeriesLong
 
-	return data.Frames{outputFrame}, nil
+	// Fill rows
+	for i, row := range sortedRows {
+		timeCol[i] = row.entry.t
+		for _, name := range valueNames {
+			if val, ok := row.entry.values[name]; ok {
+				valueCols[name][i] = val
+			}
+		}
+		for _, k := range labelKeys {
+			labelCols[k][i] = row.entry.labels[k]
```

Comment:
I'm wondering why we have an `if` check for the `value` being present for a `name`, but we don't have one for the label-value being present for a label-key. I feel like label-values are just as likely to be missing across all label-keys, as are values across all value-names.